### PR TITLE
feat: add MPT implementation

### PR DESCRIPTION
**Pull request by x-mass**

### CODE REVIEW

File: libs/nil_core/CMakeLists.txt [hunk 1]
```diff
@@ -12,13 +12,15 @@ if(CMAKE_CXX_COMPILER_VERSION VERSION_LESS 13.0)
...
-target_compile_features(${LIBRARY_NAME} INTERFACE cxx_std_23)
+add_subdirectory(src)
 
-target_include_directories(${LIBRARY_NAME} INTERFACE ${CMAKE_CURRENT_SOURCE_DIR}/include)
+target_compile_features(${LIBRARY_NAME} PUBLIC cxx_std_23)
 
-target_link_libraries(${LIBRARY_NAME} INTERFACE intx::intx sszpp::sszpp)
+target_include_directories(${LIBRARY_NAME} PUBLIC ${CMAKE_CURRENT_SOURCE_DIR}/include)
+
+target_link_libraries(${LIBRARY_NAME} PRIVATE intx::intx sszpp::sszpp)
```
Comment (makxenov) [inline libs/nil_core/CMakeLists.txt:23]: Both `intx` and `sszpp` are used in public headers, so please change scope to `PUBLIC` here

File: libs/nil_core/src/mpt/path.cpp [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+            for (int i = 0; i < least_len; ++i) {
+                if (operator[](i) != other[i]) {
+                    break;
+                }
+                common_len += 1;
+            }
+            return ConstructFromPrefix(common_len);
+        }
+
+        // Path Path::Combine(const Path& other) const { return *this + other; }
```
Comment (makxenov) [inline libs/nil_core/src/mpt/path.cpp:72]: It is unclear if it is a comment to the following method or commented code

File: libs/nil_core/include/zkevm_framework/core/mpt/mpt.hpp [hunk 3]
```diff
@@ -1,9 +1,64 @@
...
+            Node GetFromStorage(const Reference& ref) const;
+            Reference PutToStorage(const Node& node);
+            std::optional<std::vector<std::byte>> GetNode(const Reference& nodeRef,
+                                                          Path& path) const;
+            Reference SetNode(const Reference& nodeRef, Path& path,
+                              const std::vector<std::byte>& value);
+            details::DeleteResult DeleteNode(const Reference& nodeRef, const Path& path);
+
+          private:
+            Path PathFromKey(const std::vector<std::byte>& key) const;
```
Comment (makxenov) [inline libs/nil_core/include/zkevm_framework/core/mpt/mpt.hpp:50]: Please make this method static

File: libs/nil_core/include/zkevm_framework/core/mpt/path.hpp [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+            std::byte at(std::size_t idx) const;
+
+            bool StartsWith(const Path &other) const;
+            Path *Consume(std::size_t amount);
+            Path CommonPrefix(const Path &other) const;
+
+            // Methods used by sszpp. NO hash_tree_root METHOD PROVIDED
+            constexpr std::size_t ssz_size() const noexcept { return 1 + size() / 2; }
+
+            constexpr void serialize(ssz::ssz_iterator auto result) const {
```
[the remark this answers is not in the document]
Comment (x-mass) [reply libs/nil_core/include/zkevm_framework/core/mpt/path.hpp:32]: It has to be `constexpr`, thus, we can't move it to source file

**Review by akokoshn — APPROVED** [review]

**Pull request merged**